diff-models/core/src/model.ts

```diff
@@ -0,0 +1,635 @@
+import _ from 'underscore';
+import { BlockHistory } from './properties.js';
+import cloneDeep from 'clone-deep';
+import { strict as assert } from 'node:assert';
+
+/**
+ * This model may need updating pending
+ * https://github.com/cosmos/ibc/issues/825 (model updated, spec has open PR)
+ * https://github.com/cosmos/ibc/issues/796 (model updated, spec awaiting PR)
+ * TODO: implement automatic commit hash comparison warning against spec
+ */
+
+import {
+  P,
+  C,
+  UNBONDING_SECONDS_P,
+  UNBONDING_SECONDS_C,
+  NUM_VALIDATORS,
+  MAX_VALIDATORS,
+  JAIL_SECONDS,
+  BLOCK_SECONDS,
+  Event,
+} from './constants.js';
+
+import {
+  Undelegation,
+  Unval,
+  Vsc,
+  VscMatured,
+  Packet,
+  Chain,
+  Validator,
+  PacketData,
+  Slash,
+  InvariantSnapshot,
+  Status,
+  ModelInitState,
+} from './common.js';
+
+/**
+ * Store outbound packets in FIFO order from a given chain.
+ * The number of block commits for each packet is stored,
+ * and deliverable packets can be consumed once they are sufficiently
+ * committed. This mimics real IBC connections.
+ */
+class Outbox {
+  model;
+  chain;
+  // [packet, num commits]
+  fifo: [Packet, number][];
+
+  constructor(model: Model, chain: Chain) {
+    this.model = model;
+    this.chain = chain;
+    this.fifo = [];
+  }
+
+  /**
+   * Adds a packet to the outbox, with 0 commits.
+   * @param data packet data
+   */
+  add = (data: PacketData) => {
+    this.fifo.push([{ data, sendHeight: this.model.h[this.chain] }, 0]);
+  };
+
+  /**
+   * Get and internally delete deliverable packets from the outbox.
+   * @param num max num packets to consume
+   * @returns A list of deliverable packets
+   */
+  consume = (num: number): Packet[] => {
+    const [available, unavailable] = _.partition(
+      this.fifo,
+      (e) => 1 < e[1],
+    );
+    const take = available.slice(0, num);
+    this.fifo = available.slice(num).concat(unavailable);
+    return take.map((e) => e[0]);
+  };
+
+  /**
+   * Commit the packets in the outbox. Once a packet has been
+   * committed twice it is available for delivery, as per the
+   * ibc light-client functioning.
+   */
+  commit = () => {
+    this.fifo = this.fifo.map((e) => [e[0], e[1] + 1]);
+  };
+}
+
+class Staking {
+  // Model handle
+  m;
+  // Validator delegations from the delegator
+  // A fixed descending order is used for the initial values to allow
+  // easy setup in the SUT.
+  delegation: number[];
```

Comment:
Good question, I should definitely document this better.

Delegation is the delegation from the modelled delegator to the validator. Tokens is the total voting power of the validator. The voting power <-> token ratio is 1:1. There are more Tokens than Delegation because there is _ghost_ delegation. This is extra delegation that each validator has to stop it being deleted. 

This is needed because the SUT has a concept of minSelfDelegation. So this difference (tokens[i]-delegation[i]) is effectively minSelfDelegation

Reply:
I agree it's not clear